### ISSUE TITLE
CUDA validators crashlooping while other cuda containers run fine

### ISSUE THREAD

**Issue by faiq** [NVIDIA/gpu-operator#389]
_The template below is mostly useful for bug reports and support questions. Feel free to remove anything which doesn't apply to you and add more information where it makes sense._

### 1. Quick Debug Checklist
- [ ] Are you running on an Ubuntu 18.04 node?
- [x] Are you running Kubernetes v1.13+?
- [ ] Are you running Docker (>= 18.06) or CRIO (>= 1.13+)?
- [ ] Do you have `i2c_core` and `ipmi_msghandler` loaded on the nodes?
- [ ] Did you apply the CRD (`kubectl describe clusterpolicies --all-namespaces`)

### 1. Issue or feature description:

The cuda validators in my cluster are in `Init:CrashLoopBackOff `while other cuda vector addition workloads run completely fine. 

```sh 
$ kubectl get po 
NAME                                       READY   STATUS                  RESTARTS        AGE
gpu-feature-discovery-cqjqx                1/1     Running                 0               28m
gpu-feature-discovery-x4qh2                1/1     Running                 0               28m
gpu-operator-77787587cf-cxnzl              1/1     Running                 0               28m
nvidia-container-toolkit-daemonset-bk28j   1/1     Running                 0               28m
nvidia-container-toolkit-daemonset-qvftc   1/1     Running                 0               28m
nvidia-cuda-validator-ccn65                0/1     Init:CrashLoopBackOff   5 (23s ago)     3m29s
nvidia-cuda-validator-p7sgd                0/1     Init:CrashLoopBackOff   5 (16s ago)     3m18s
nvidia-dcgm-exporter-p5wrc                 1/1     Running                 0               28m
nvidia-dcgm-exporter-rvnz6                 1/1     Running                 0               28m
nvidia-device-plugin-daemonset-2bfmt       1/1     Running                 0               28m
nvidia-device-plugin-daemonset-pvphw       1/1     Running                 0               28m
nvidia-operator-validator-5qfx2            0/1     Init:2/4                5 (4m55s ago)   28m
nvidia-operator-validator-nc9n6            0/1     Init:2/4                5 (4m40s ago)   28m
```

On closer inspection its the vector add that's giving us an issue when getting the logs we get

```
$ kubectl logs nvidia-cuda-validator-ccn65 -c cuda-validation --previous 
Failed to allocate device vector A (error code all CUDA-capable devices are busy or unavailable)!
[Vector addition of 50000 elements]
``` 
However, I am able to deploy other cudavector workloads for example 

```
$ kubectl get po -n e2e-gpu-workload 
NAME                READY   STATUS             RESTARTS         AGE
cuda-vector-add     0/1     Completed          0                29m
cuda-vector-add-2   0/1     CrashLoopBackOff   10 (2m19s ago)   14m
cuda-vector-add-3   0/1     CrashLoopBackOff   6 (4m19s ago)    11m
cuda-vector-add-4   0/1     Completed          0                9m34s
```

and another look, printing out the container images for the running and non-running pods 

```
$ kubectl get po -n e2e-gpu-workload -o custom-columns=CONTAINER:.spec.containers[0].name,IMAGE:.spec.containers[0].image
CONTAINER        IMAGE
cuda-vectoradd   nvidia/samples:vectoradd-cuda11.2.1
cuda-vectoradd   nvcr.io/nvidia/cloud-native/gpu-operator-validator:v1.11.1
cuda-vectoradd   docker.io/anjia0532/cuda-vector-add:v0.1
cuda-vectoradd   docker.io/anjia0532/cuda-vector-add:v0.1
```
when we look at the logs we find that everything works fine 

```
$ kubectl logs -n e2e-gpu-workload cuda-vector-add
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done
``` 
and expanding the pod definition we get 

```yaml 
$ kubectl get po -n e2e-gpu-workload cuda-vector-add -o yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    cni.projectcalico.org/containerID: 9cbccd5849cd9f8a0b1670b84392deacfac4a55eaa05541ef83ab66df24de089
    cni.projectcalico.org/podIP: ""
    cni.projectcalico.org/podIPs: ""
  creationTimestamp: "2022-08-10T22:59:38Z"
  name: cuda-vector-add
  namespace: e2e-gpu-workload
  resourceVersion: "16028"
  uid: 17064311-b8a5-4ff3-bd4c-c3a9665d2ec4
spec:
  containers:
  - image: nvidia/samples:vectoradd-cuda11.2.1
    imagePullPolicy: IfNotPresent
    name: cuda-vectoradd
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-d7zbx
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  nodeSelector:
    nvidia.com/gpu.count: "1"
```

### 2. Steps to reproduce the issue

I installed the operator with the following command:

```
helm install --wait --generate-name  --set nfd.enabled=false --set driver.enabled=false --set toolkit.version=1.6.0-centos7  nvidia/gpu-operator
``` 

In my setup I installed the nvidia device drivers and cuda drivers on the host using the runfile and am using the operator to install the container runtime. 

### 3. Information to [attach](https://help.github.com/articles/file-attachments-on-issues-and-pull-requests/) (optional if deemed irrelevant)

 - [ ] kubernetes pods status: `kubectl get pods --all-namespaces`
 - [ ] kubernetes daemonset status: `kubectl get ds --all-namespaces`
 - [ ] If a pod/ds is in an error state or pending state `kubectl describe pod -n NAMESPACE POD_NAME`
 - [ ] If a pod/ds is in an error state or pending state `kubectl logs -n NAMESPACE POD_NAME`

 - [ ] Output of running a container on the GPU machine: `docker run -it alpine echo foo`
 - [ ] Docker configuration file: `cat /etc/docker/daemon.json`
 - [ ] Docker runtime configuration: `docker info | grep runtime`

 - [ ] NVIDIA shared directory: `ls -la /run/nvidia`
 - [x] NVIDIA packages directory: `ls -la /usr/local/nvidia/toolkit`
 - [ ] NVIDIA driver directory: `ls -la /run/nvidia/driver`
 - [ ] kubelet logs `journalctl -u kubelet > kubelet.logs`

When I run nvidia-smi on the host i get the following, indicating that cuda is installed alright.

```
[root@ip-10-0-101-7 bin]# nvidia-smi
Wed Aug 10 23:40:28 2022
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.82.01    Driver Version: 470.82.01    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla K80           Off  | 00000000:00:1E.0 Off |                    0 |
| N/A   35C    P8    30W / 149W |      0MiB / 11441MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

as well as running  a container with cuda 

```
[root@ip-10-0-101-7 bin]# ctr run --rm --gpus 0 -t docker.io/nvidia/samples:vectoradd-cuda11.2.1 add2
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done
```

```
[root@ip-10-0-101-7 bin]# ls -la /usr/local/nvidia/toolkit
total 8548
drwxr-xr-x. 3 root root    4096 Aug 10 23:04 .
drwxr-xr-x. 3 root root      21 Aug 10 23:04 ..
drwxr-xr-x. 3 root root      38 Aug 10 23:04 .config
lrwxrwxrwx. 1 root root      28 Aug 10 23:04 libnvidia-container.so.1 -> libnvidia-container.so.1.4.0
-rwxr-xr-x. 1 root root  179192 Aug 10 23:04 libnvidia-container.so.1.4.0
-rwxr-xr-x. 1 root root     154 Aug 10 23:04 nvidia-container-cli
-rwxr-xr-x. 1 root root   43024 Aug 10 23:04 nvidia-container-cli.real
-rwxr-xr-x. 1 root root     342 Aug 10 23:04 nvidia-container-runtime
-rwxr-xr-x. 1 root root     350 Aug 10 23:04 nvidia-container-runtime-experimental
-rwxr-xr-x. 1 root root 3991000 Aug 10 23:04 nvidia-container-runtime.experimental
lrwxrwxrwx. 1 root root      24 Aug 10 23:04 nvidia-container-runtime-hook -> nvidia-container-toolkit
-rwxr-xr-x. 1 root root 2359384 Aug 10 23:04 nvidia-container-runtime.real
-rwxr-xr-x. 1 root root     198 Aug 10 23:04 nvidia-container-toolkit
-rwxr-xr-x. 1 root root 2147896 Aug 10 23:04 nvidia-container-toolkit.real
[root@ip-10-0-101-7 bin]#
```

Any guidance would be greatly appreciated, and please let me know how I can help. 

Thank you.

**Comment by shivamerla** [comment]
@faiq I see that container-toolkit version is very old (1.4.0) where Operator v1.11.1 deploys toolkit version 1.10.0. Wondering if this is causing the issue as vectorAdd sample we use is latest one based on CUDA 11.7. Can you update the container toolkit to more latest one?

**Comment by faiq** [comment]
I can absolutely give that a try with the adjusted value.

In a related topic, would you be interested in a contribution to infer the toolkit os in a similar way as the operator does for driver does for driver. As there seems to be OS specific library dependencies see: https://github.com/NVIDIA/gpu-operator/issues/72#issuecomment-742023528

IMO this would greatly improve the UX of the operator.

**Comment by shivamerla** [comment]
@faiq yes, that was the plan initially but we couldn't do that as toolkit  images were not available for all variants. Also, toolkit image with OS variant `ubuntu18.04` works across `ubuntu20.04` and `ubuntu22.04`. Also, once the fix suggested by Kevin in the linked thread is complete, we don't need different variants at all for each version of a given OS.

**Comment by BHSDuncan** [comment]
I'm having a very similar issue when installing the Cloud Native Core (6.2) stack via one of the available Ansible playbooks. However, my container toolkit is 1.10.0. Yet I still get the same error message as OP.

I can give the rest of the versions of whatever is needed here, but I wanted to check if it's ok to do so here, or if I should start a new issue (and if so, should it be here or in the CNC repository)?

Thank you.

**Comment by faiq** [comment]
@shivamerla I've tried again with a newer version of the container toolkit - v1.10.0-centos-7 and still no luck 

```yaml 
$ kubectl describe po nvidia-container-toolkit-daemonset-hb5z6
Name:                 nvidia-container-toolkit-daemonset-hb5z6
Namespace:            default
Priority:             2000001000
Priority Class Name:  system-node-critical
Node:                 ip-10-0-115-55.us-west-2.compute.internal/10.0.115.55
Start Time:           Fri, 12 Aug 2022 09:33:01 -0700
Labels:               app=nvidia-container-toolkit-daemonset
                      controller-revision-hash=bd5d94876
                      pod-template-generation=2
Annotations:          cni.projectcalico.org/containerID: 54adee8127a6836a1ce81df53b3a8d9c8bcd7c4f61d696c4b7b5f36263221ca3
                      cni.projectcalico.org/podIP: 192.168.114.134/32
                      cni.projectcalico.org/podIPs: 192.168.114.134/32
Status:               Running
IP:                   192.168.114.134
IPs:
  IP:           192.168.114.134
Controlled By:  DaemonSet/nvidia-container-toolkit-daemonset
Init Containers:
  driver-validation:
    Container ID:  containerd://179a13f5197da79feac867a2498529f102e12f92d283f273ae430581b6d8e81c
    Image:         nvcr.io/nvidia/cloud-native/gpu-operator-validator:v1.11.0
    Image ID:      nvcr.io/nvidia/cloud-native/gpu-operator-validator@sha256:d83aee1c7a9c3aa39e3947f0bd206e7955390943b8e3f2e56bdb96bc4ef99d69
    Port:          <none>
    Host Port:     <none>
    Command:
      sh
      -c
    Args:
      nvidia-validator
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Fri, 12 Aug 2022 09:33:02 -0700
      Finished:     Fri, 12 Aug 2022 09:33:04 -0700
    Ready:          True
    Restart Count:  0
    Environment:
      WITH_WAIT:  true
      COMPONENT:  driver
    Mounts:
      /host from host-root (ro)
      /run/nvidia/driver from driver-install-path (rw)
      /run/nvidia/validations from run-nvidia-validations (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-njrtz (ro)
Containers:
  nvidia-container-toolkit-ctr:
    Container ID:  containerd://05d50be00b768266ce5f603505e9cae580888ca4a1fbeb076835e80b94b649d7
    Image:         nvcr.io/nvidia/k8s/container-toolkit:v1.10.0-centos7
    Image ID:      nvcr.io/nvidia/k8s/container-toolkit@sha256:80c05e1db6462d6561cfa46d9d0c58bb0e9c222df8bcada56182b3a58c48072c
    Port:          <none>
    Host Port:     <none>
    Command:
      bash
      -c
    Args:
      [[ -f /run/nvidia/validations/host-driver-ready ]] && driver_root=/ || driver_root=/run/nvidia/driver; export NVIDIA_DRIVER_ROOT=$driver_root; exec nvidia-toolkit /usr/local/nvidia
    State:          Running
      Started:      Fri, 12 Aug 2022 09:33:37 -0700
    Ready:          True
    Restart Count:  0
    Environment:
      RUNTIME_ARGS:              --socket /runtime/sock-dir/containerd.sock --config /runtime/config-dir/config.toml
      RUNTIME:                   containerd
      CONTAINERD_RUNTIME_CLASS:  nvidia
    Mounts:
      /host from host-root (ro)
      /run/nvidia from nvidia-run-path (rw)
      /runtime/config-dir/ from containerd-config (rw)
      /runtime/sock-dir/ from containerd-socket (rw)
      /usr/local/nvidia from nvidia-local (rw)
      /usr/share/containers/oci/hooks.d from crio-hooks (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-njrtz (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  nvidia-run-path:
    Type:          HostPath (bare host directory volume)
    Path:          /run/nvidia
    HostPathType:  DirectoryOrCreate
  run-nvidia-validations:
    Type:          HostPath (bare host directory volume)
    Path:          /run/nvidia/validations
    HostPathType:  DirectoryOrCreate
  driver-install-path:
    Type:          HostPath (bare host directory volume)
    Path:          /run/nvidia/driver
    HostPathType:  
  host-root:
    Type:          HostPath (bare host directory volume)
    Path:          /
    HostPathType:  
  nvidia-local:
    Type:          HostPath (bare host directory volume)
    Path:          /usr/local/nvidia
    HostPathType:  
  crio-hooks:
    Type:          HostPath (bare host directory volume)
    Path:          /run/containers/oci/hooks.d
    HostPathType:  
  containerd-config:
    Type:          HostPath (bare host directory volume)
    Path:          /etc/containerd
    HostPathType:  
  containerd-socket:
    Type:          HostPath (bare host directory volume)
    Path:          /run/containerd
    HostPathType:  
  kube-api-access-njrtz:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              nvidia.com/gpu.deploy.container-toolkit=true
Tolerations:                 node.kubernetes.io/disk-pressure:NoSchedule op=Exists
                             node.kubernetes.io/memory-pressure:NoSchedule op=Exists
                             node.kubernetes.io/not-ready:NoExecute op=Exists
                             node.kubernetes.io/pid-pressure:NoSchedule op=Exists
                             node.kubernetes.io/unreachable:NoExecute op=Exists
                             node.kubernetes.io/unschedulable:NoSchedule op=Exists
                             nvidia.com/gpu:NoSchedule op=Exists
Events:
```

Failing pod 

```
nvidia-cuda-validator-dgmwl                0/1     Init:CrashLoopBackOff   2 (12s ago)   30s
```

failing pod logs 

```
$ kubectl logs nvidia-cuda-validator-dgmwl --previous -c cuda-validation
[Vector addition of 50000 elements]
Failed to allocate device vector A (error code all CUDA-capable devices are busy or unavailable)!
``` 

Completed pod + image

```
 kubectl get po cuda-vector-add -o yaml | grep image
      {"apiVersion":"v1","kind":"Pod","metadata":{"annotations":{},"name":"cuda-vector-add","namespace":"default"},"spec":{"containers":[{"image":"nvidia/samples:vectoradd-cuda11.2.1","name":"cuda-vectoradd"}],"nodeSelector":{"nvidia.com/gpu.count":"1"},"restartPolicy":"OnFailure"}}
  - image: nvidia/samples:vectoradd-cuda11.2.1
    imagePullPolicy: IfNotPresent
    image: docker.io/nvidia/samples:vectoradd-cuda11.2.1
``` 

I'll try with this new RC cut and see if it resolves the issue, but it seems like its a problem with the test itself - v1.11.0-rc.2-centos7

**Comment by shivamerla** [comment]
@elezar Any ideas what might cause this, we are using cuda sample from `nvcr.io/nvidia/k8s/cuda-sample:vectoradd-cuda11.6.0-ubi8`

**Comment by shivamerla** [comment]
@faiq in your example pod, can you use this image and confirm? `nvcr.io/nvidia/k8s/cuda-sample:vectoradd-cuda11.6.0-ubi8`

**Comment by faiq** [comment]
It looks like it was an issue with the vecrtor-add pod 

```
$ kubectl logs cuda-vector-add-11-6
[Vector addition of 50000 elements]
Failed to allocate device vector A (error code CUDA driver version is insufficient for CUDA runtime version)!
``` 

```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    cni.projectcalico.org/containerID: d5597e5f0e96635b3c6b883dc2ec65c65aa3480d0ee61778d2e7b11ddde3f5fe
    cni.projectcalico.org/podIP: 192.168.114.147/32
    cni.projectcalico.org/podIPs: 192.168.114.147/32
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Pod","metadata":{"annotations":{},"name":"cuda-vector-add-11-6","namespace":"default"},"spec":{"containers":[{"image":"nvidia/samples:vectoradd-cuda11.6.0","name":"cuda-vectoradd"}],"nodeSelector":{"nvidia.com/gpu.count":"1"},"restartPolicy":"OnFailure"}}
  creationTimestamp: "2022-08-12T16:56:41Z"
  name: cuda-vector-add-11-6
  namespace: default
  resourceVersion: "28397"
  uid: 0ad82d59-413c-44a8-b416-9ee71508220f
spec:
  containers:
  - image: nvidia/samples:vectoradd-cuda11.6.0
    imagePullPolicy: IfNotPresent
    name: cuda-vectoradd
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-mnzft
      readOnly: true
```

**Comment by faiq** [comment]
The ubi8 image didn't give me much luck either 

```
$ kubectl get po  cuda-vector-add-11-6-ubi8 -o yaml | yq r - status.containerStatuses
- containerID: containerd://6ef72f13361ca41084180ebc5c25ff18ac9bcb2930a4dd995c255e7db4aa6ea2
  image: nvcr.io/nvidia/k8s/cuda-sample:vectoradd-cuda11.6.0-ubi8
  imageID: nvcr.io/nvidia/k8s/cuda-sample@sha256:610e53e5e2dc543d5188db6d9dea3ca3829fdf6b0ae93edfcf1012a7ecc9a58e
  lastState:
    terminated:
      containerID: containerd://6ef72f13361ca41084180ebc5c25ff18ac9bcb2930a4dd995c255e7db4aa6ea2
      exitCode: 1
      finishedAt: "2022-08-12T17:01:25Z"
      reason: Error
      startedAt: "2022-08-12T17:01:25Z"
  name: cuda-vectoradd
  ready: false
  restartCount: 4
  started: false
  state:
    waiting:
      message: back-off 1m20s restarting failed container=cuda-vectoradd pod=cuda-vector-add-11-6-ubi8_default(3b8bb11a-d849-4255-ac4a-b2c6e4ca5233)
      reason: CrashLoopBackOff

```

**Comment by shivamerla** [comment]
I will try to reproduce, @faiq can you provide the exact config details?

OS Version
GPU Operator Version
Driver Pre-installed?
Container-Toolkit Pre-installed?
Container-Toolkit Version

**Comment by faiq** [comment]
OS Version: centos-7.9 
GPU Operator Version
Driver Pre-installed: yes
Container-Toolkit Pre-installed: no
Container-Toolkit Version:  nvcr.io/nvidia/k8s/container-toolkit:v1.10.0-centos7
GPU Type: tesla k80 

installed drivers

```
[root@ip-10-0-115-55 bin]# nvidia-smi
Fri Aug 12 17:27:02 2022
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.82.01    Driver Version: 470.82.01    CUDA Version: 11.4     |
```

**Comment by shivamerla** [comment]
@faiq vectorAdd is dependent on libcuda.so.1 which is injected by the toolkit with underlying driver version.


`kubectl run vector-add --image=nvcr.io/nvidia/cloud-native/gpu-operator-validator:v1.11.1 --command -- /bin/bash -c "sleep inf"`

`kubectl  exec -it vector-add -- /bin/bash`

```
[root@vector-add /]# ls -ltr ./usr/lib64/ | grep libcuda
-rwxr-xr-x  1 root root 22851528 Jan 25  2022 libcuda.so.510.47.03
-rwxr-xr-x  1 root root 20988000 Aug 12 11:14 libcuda.so.515.65.01
lrwxrwxrwx  1 root root       12 Aug 12 17:52 libcuda.so -> libcuda.so.1
lrwxrwxrwx  1 root root       20 Aug 12 17:52 libcuda.so.1 -> libcuda.so.515.65.01

[root@vector-add /]# vectorAdd
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done
```

Can you confirm in your case libraries for 470.82.01 driver are being injected correctly?

**Comment by faiq** [comment]
Well bingo

```
$ kubectl get po
NAME                                       READY   STATUS                  RESTARTS         AGE
cuda-vector-add                            0/1     Completed               0                87m
cuda-vector-add-11-6                       0/1     CrashLoopBackOff        18 (42s ago)     68m
cuda-vector-add-11-6-ubi8                  0/1     CrashLoopBackOff        17 (3m10s ago)   65m
cuda-vector-add-2                          0/1     CrashLoopBackOff        19 (4m14s ago)   76m
gpu-feature-discovery-ccjwq                1/1     Running                 0                97m
gpu-operator-77787587cf-wfwtt              1/1     Running                 0                97m
nvidia-container-toolkit-daemonset-xp2hv   1/1     Running                 0                77m
nvidia-cuda-validator-lmkx4                0/1     Init:CrashLoopBackOff   5 (58s ago)      3m52s
nvidia-dcgm-exporter-b7hq8                 1/1     Running                 0                97m
nvidia-device-plugin-daemonset-mzrs4       1/1     Running                 0                97m
nvidia-operator-validator-jrf5p            0/1     Init:2/4                13 (8m59s ago)   97m
nvidia-version-check                       0/1     CrashLoopBackOff        10 (2m10s ago)   28m
vector-add                                 1/1     Running                 0                19s

11:04 AM  faiqraza @ linux  ~/go/src/github.com/mesosphere/konvoy-image-builder  faiq/nvidia-runfile⚡? 
$ kubectl exec -it vector-add -- /bin/bash
[root@vector-add /]# ls -ltr ./usr/lib64/ | grep libcuda
[root@vector-add /]# 
```

My .so files are not in /usr/lib64 I can make a link and try again.

**Comment by faiq** [comment]
Again no luck.

On the host 

```
[root@ip-10-0-115-55 bin]# ls /usr/lib64/| grep libcuda
libcuda.so
libcuda.so.1
libcuda.so.470.82.01
```

```
$ kubectl exec -it vector-add3 -- /bin/bash
[root@vector-add3 /]# ls -ltr /usr/lib64/| grep libcuda
[root@vector-add3 /]# ls -ltr /usr/lib64/| grep nvidia
[root@vector-add3 /]# 
```

**Comment by shivamerla** [comment]
Ok, need to debug more from toolkit side to understand whey these are not injected. `cat /usr/local/nvidia/toolkit/.config/nvidia-container-runtime/config.toml` and also add below debug fields.

```
disable-require = false

[nvidia-container-cli]
  environment = []
  ldconfig = "@/run/nvidia/driver/sbin/ldconfig.real"
  load-kmods = true
  path = "/usr/local/nvidia/toolkit/nvidia-container-cli"
  root = "/"
  debug = "/var/log/nvidia-container-cli.log" <<<<<<<<<<<<<

[nvidia-container-runtime]
  log-level = "info"
  debug = "/var/log/nvidia-container-runtime.log" <<<<<<<<<<<<<<
  mode = "auto"
  runtimes = ["docker-runc", "runc"]

  [nvidia-container-runtime.modes]

    [nvidia-container-runtime.modes.csv]
      mount-spec-path = "/etc/nvidia-container-runtime/host-files-for-container.d"
```
After this, when you run the vectorAdd sample again, these log files will be generated on the host. We can debug using the logs to dig further.

**Comment by faiq** [comment]
From the host 

```
[root@ip-10-0-115-55 bin]# cat /usr/local/nvidia/toolkit/.config/nvidia-container-runtime/config.toml
accept-nvidia-visible-devices-as-volume-mounts = false
accept-nvidia-visible-devices-envvar-when-unprivileged = false
disable-require = false

[nvidia-container-cli]
  environment = []
  ldconfig = "@/sbin/ldconfig"
  load-kmods = true
  path = "/usr/local/nvidia/toolkit/nvidia-container-cli"
  root = "/"

[nvidia-container-runtime]
  log-level = "info"
  mode = "auto"
  runtimes = ["docker-runc", "runc"]

  [nvidia-container-runtime.modes]

    [nvidia-container-runtime.modes.csv]
      mount-spec-path = "/etc/nvidia-container-runtime/host-files-for-container.d"
```

**Comment by shivamerla** [comment]
I think it is caused by `accept-nvidia-visible-devices-envvar-when-unprivileged = false` setting. I recommend to un-install operator, wait for all pods to be deleted, cleanup /usr/local/nvidia/toolkit folder and re-install operator. With toolkit version `v1.10.0` this won't happen. Did you manually add these settings?

**Comment by faiq** [comment]
I did not manually add these settings. I'll try installing it again.

**Comment by faiq** [comment]
I reinstalled and now I'm getting the libcuda.so files in the container

```
[root@nvidia-version-check-2 /]# ls /usr/lib64/ | grep cuda
libcuda.so
libcuda.so.1
libcuda.so.470.82.01
libcuda.so.510.47.03
``` 

Not sure why i have this 510.47.03 in there as well.

**Comment by faiq** [comment]
Also within the container i can run nvidia-smi

```
[root@nvidia-version-check-2 /]# nvidia-smi
Fri Aug 12 20:57:36 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.82.01    Driver Version: 470.82.01    CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla K80           Off  | 00000000:00:1E.0 Off |                    0 |
| N/A   34C    P8    27W / 149W |      0MiB / 11441MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
[root@nvidia-version-check-2 /]# 
```
but the vectorAdd program still does not work 

```
[root@nvidia-version-check-2 /]# vectorAdd 
[Vector addition of 50000 elements]
Failed to allocate device vector A (error code all CUDA-capable devices are busy or unavailable)!
[root@nvidia-version-check-2 /]# 

Also my host nvidia-smi is giving me a different result for CUDA 

```
sh-4.2$ bash
[ssm-user@ip-10-0-101-96 bin]$ nvidia-smi
Fri Aug 12 21:00:05 2022
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.82.01    Driver Version: 470.82.01    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla K80           Off  | 00000000:00:1E.0 Off |                    0 |
| N/A   34C    P8    26W / 149W |      0MiB / 11441MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
[ssm-user@ip-10-0-101-96 bin]$
```

**Comment by faiq** [comment]
Also the results of 

```
[ssm-user@ip-10-0-101-96 bin]$  cat /usr/local/nvidia/toolkit/.config/nvidia-container-runtime/config.toml
disable-require = false

[nvidia-container-cli]
  environment = []
  ldconfig = "@/sbin/ldconfig"
  load-kmods = true
  path = "/usr/local/nvidia/toolkit/nvidia-container-cli"
  root = "/"

[nvidia-container-runtime]
  log-level = "info"
  mode = "auto"
  runtimes = ["docker-runc", "runc"]

  [nvidia-container-runtime.modes]

    [nvidia-container-runtime.modes.csv]
      mount-spec-path = "/etc/nvidia-container-runtime/host-files-for-container.d"
[ssm-user@ip-10-0-101-96 bin]$
```

there no longer is 
`accept-nvidia-visible-devices-envvar-when-unprivileged = false`

**Comment by faiq** [comment]
It really looks like the image to me for the validation. 

When using this pod spec with `nvidia/samples:vectoradd-cuda11.2.1` everything works as expected. 

```
$ cat check.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: nvidia-version-check-2
spec:
  restartPolicy: OnFailure
  containers:
  - name: nvidia-version-check
    #image: "nvcr.io/nvidia/cloud-native/gpu-operator-validator:v1.11.1"
    image: "nvidia/samples:vectoradd-cuda11.2.1"
    command: ["sleep"]
    args: ["inf"]
    resources:
      limits:
         nvidia.com/gpu: "1"

```
 
 
The results of nvidia-smi are no longer incorrect with this image and I can use vector add 

```
root@nvidia-version-check-2:/# nvidia-smi 
Fri Aug 12 21:20:46 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.82.01    Driver Version: 470.82.01    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla K80           Off  | 00000000:00:1E.0 Off |                    0 |
| N/A   35C    P8    26W / 149W |      0MiB / 11441MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
root@nvidia-version-check-2:/# /tmp/sample 
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done
root@nvidia-version-check-2:/# /tmp/vectorAdd 
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done
```

**Comment by shivamerla** [comment]
thanks @faiq will check with container-toolkit folks on cause for this and update next week.

**Comment by BHSDuncan** [comment]
Could this also be causing what looks to be a very similar issue when using the Cloud Native Core Ansible playbook (6.2)?

**Comment by shivamerla** [comment]
@BHSDuncan Is your config with CentOS 7 too? We are still debugging this issue, will update this week.

**Comment by BHSDuncan** [comment]
Ubuntu 20.04. I'm currently trying to upgrade my NVIDIA driver to a 515 version so it could be that I'm a red herring. Just FYI.

**Comment by BHSDuncan** [comment]
I can confirm that updating to the latest NVIDIA GPU driver version (515.x) has corrected this issue. I apologize for the red herring.

**Comment by shivamerla** [comment]
@faiq i have tried to recreate this with CentOS 7, but no luck with A100 cards. I have confirmed with @klueska that cuda-runtime version mismatch is expected due to CUDA forward compatibility. The compat libs included in the image are from CUDA 11.6 (R510 driver), which is being used as its forward compatible with older 11.4(R470) drivers installed. But i think the limitation here is the type of GPU. From the guide here, looks like only data center GPUs support it and in your case is Tesla K80. Need to check if there is a way to override compat with the toolkit for your case, so that 11.4 libraries will be used instead.

https://docs.nvidia.com/deploy/cuda-compatibility/index.html#forward-compatible-upgrade

Excerpt from this link:

```
Forward Compatibility is applicable only for systems with NVIDIA Data Center GPUs or select [NGC Server Ready](https://docs.nvidia.com/ngc/ngc-ready-systems/index.html#abstract) SKUs of RTX cards. It’s mainly intended to support applications built on newer CUDA Toolkits to run on systems installed with an older NVIDIA Linux GPU driver from different major release families.
```

The error codes you are seeing align with this:

```
CUDA_ERROR_SYSTEM_DRIVER_MISMATCH = 803. This error indicates that there is a mismatch between the versions of the display driver and the CUDA driver.
CUDA_ERROR_COMPAT_NOT_SUPPORTED_ON_DEVICE = 804. This error indicates that the system was upgraded to run with forward compatibility but the visible hardware detected by CUDA does not support this configuration.
```


```
[root@k8s-centos-1 ~]# cat /etc/os-release 
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

[root@k8s-centos-1 ~]# nvidia-smi
Mon Aug 15 16:21:16 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.103.01   Driver Version: 470.103.01   CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  GRID A100-8C        On   | 00000000:02:01.0 Off |                    0 |
| N/A   N/A    P0    N/A /  N/A |    656MiB /  8187MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
[root@k8s-centos-1 ~]# helm ls -n gpu-operator
NAME        	NAMESPACE   	REVISION	UPDATED                                	STATUS  	CHART               	APP VERSION
gpu-operator	gpu-operator	1       	2022-08-15 16:19:22.674169791 -0400 EDT	deployed	gpu-operator-v1.11.1	v1.11.1    
[root@k8s-centos-1 ~]# kubectl get pods -n gpu-operator
NAME                                                          READY   STATUS      RESTARTS   AGE
gpu-feature-discovery-wrtzz                                   1/1     Running     0          105s
gpu-operator-84d9f557c8-nwkg8                                 1/1     Running     0          2m3s
gpu-operator-node-feature-discovery-master-59b4b67f4f-7wtm7   1/1     Running     0          2m3s
gpu-operator-node-feature-discovery-worker-8c9qv              1/1     Running     0          2m3s
nvidia-container-toolkit-daemonset-nndfz                      1/1     Running     0          105s
nvidia-cuda-validator-nzxpv                                   0/1     Completed   0          89s
nvidia-dcgm-exporter-mkz9f                                    1/1     Running     0          105s
nvidia-device-plugin-daemonset-5rwbb                          1/1     Running     0          105s
nvidia-device-plugin-validator-6bl7l                          0/1     Completed   0          67s
nvidia-mig-manager-rnfch                                      1/1     Running     0          105s
nvidia-operator-validator-vjzzb                               1/1     Running     0          105s


[root@k8s-centos-1 ~]# kubectl  describe pod nvidia-cuda-validator-nzxpv  -n gpu-operator
Name:         nvidia-cuda-validator-nzxpv
Namespace:    gpu-operator
Priority:     0
Node:         k8s-centos-1/10.150.187.70
Start Time:   Mon, 15 Aug 2022 16:19:56 -0400
Labels:       app=nvidia-cuda-validator
Annotations:  cni.projectcalico.org/containerID: cc16539c78e01e393219b34b15e7f81100005751f54073d885fb01206f569d59
              cni.projectcalico.org/podIP: 
              cni.projectcalico.org/podIPs: 
Status:       Succeeded
IP:           192.168.24.149
IPs:
  IP:           192.168.24.149
Controlled By:  ClusterPolicy/cluster-policy
Init Containers:
  cuda-validation:
    Container ID:  containerd://d89182cae3800d98911a474a9e8a8e2835299c52ec092b28c51b09e519cfc57e
    Image:         nvcr.io/nvidia/cloud-native/gpu-operator-validator:v1.11.1
    Image ID:      nvcr.io/nvidia/cloud-native/gpu-operator-validator@sha256:10d825ccbfcf87f8c74f841fa89ae345f2d381fb47a5b2b31f61b72ebbbb0915
    Port:          <none>
    Host Port:     <none>
    Command:
      sh
      -c
    Args:
      vectorAdd
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Mon, 15 Aug 2022 16:19:56 -0400
      Finished:     Mon, 15 Aug 2022 16:19:57 -0400
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-8wltq (ro)


[root@k8s-centos-1 ~]# kubectl run vector-add --image=nvcr.io/nvidia/cloud-native/gpu-operator-validator:v1.11.1 --command -- /bin/bash -c "sleep inf"
pod/vector-add created
[root@k8s-centos-1 ~]# kubectl get pod
NAME          READY   STATUS    RESTARTS      AGE
driver-test   1/1     Running   3 (10m ago)   68d
vector-add    1/1     Running   0             2s
[root@k8s-centos-1 ~]# kubectl  exec -it vector-add -- /bin/bash
[root@vector-add /]# vectorAdd
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done
[root@vector-add /]# 
[root@vector-add /]# ls -ltr /usr/lib64/ | grep libcuda
-rwxr-xr-x  1 root root 22851528 Jan 25  2022 libcuda.so.510.47.03
-rwxr-xr-x  1 root root 24216904 Aug 15 20:09 libcuda.so.470.103.01
lrwxrwxrwx  1 root root       12 Aug 15 20:24 libcuda.so -> libcuda.so.1
lrwxrwxrwx  1 root root       20 Aug 15 20:24 libcuda.so.1 -> libcuda.so.510.47.03

```

**Comment by faiq** [comment]
@shivamerla it looks like the suggestion you gave in #391 worked i was able to get my validator image to work by removing the compat libs

```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    cni.projectcalico.org/containerID: 684414b3c939cf65e1f48c78bd7d908231a52e51d97d4ce460613644c5bf2336
    cni.projectcalico.org/podIP: ""
    cni.projectcalico.org/podIPs: ""
  creationTimestamp: "2022-08-16T21:24:14Z"
  generateName: nvidia-cuda-validator-
  labels:
    app: nvidia-cuda-validator
  name: nvidia-cuda-validator-kwhd7
  namespace: default
  ownerReferences:
  - apiVersion: nvidia.com/v1
    blockOwnerDeletion: true
    controller: true
    kind: ClusterPolicy
    name: cluster-policy
    uid: 31f3f689-d3b4-4e6f-a839-74d1ee9baf69
  resourceVersion: "92351"
  uid: 7763960b-1d58-4499-b71c-6bc5c4efefd4
spec:
  containers:
  - args:
    - echo cuda workload validation is successful
    command:
    - sh
    - -c
    image: docker.io/faiq/gpu-operator-validator:v1.11.0
    imagePullPolicy: IfNotPresent
    name: nvidia-cuda-validator
    resources: {}
    securityContext:
      allowPrivilegeEscalation: false
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-b8f59
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  initContainers:
  - args:
    - vectorAdd
    command:
    - sh
    - -c
    image: docker.io/faiq/gpu-operator-validator:v1.11.0
    imagePullPolicy: IfNotPresent
    name: cuda-validation
    resources: {}
    securityContext:
      allowPrivilegeEscalation: false
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-b8f59
      readOnly: true
  nodeName: ip-10-0-113-252.us-west-2.compute.internal
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: OnFailure
  runtimeClassName: nvidia
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: nvidia-operator-validator
  serviceAccountName: nvidia-operator-validator
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoSchedule
    key: nvidia.com/gpu
    operator: Exists
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - name: kube-api-access-b8f59
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-root-ca.crt
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2022-08-16T21:24:16Z"
    reason: PodCompleted
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2022-08-16T21:24:14Z"
    reason: PodCompleted
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2022-08-16T21:24:14Z"
    reason: PodCompleted
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2022-08-16T21:24:14Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: containerd://784a53092f791b414e708bc9ec22d86e5a92a3988a86910d28fda8f06b7e250a
    image: docker.io/faiq/gpu-operator-validator:v1.11.0
    imageID: docker.io/faiq/gpu-operator-validator@sha256:0ec2a74daaaea3717a95d9399cbc6e9486c38f60e9b54b826c7031e3c2d52dfa
    lastState: {}
    name: nvidia-cuda-validator
    ready: false
    restartCount: 0
    started: false
    state:
      terminated:
        containerID: containerd://784a53092f791b414e708bc9ec22d86e5a92a3988a86910d28fda8f06b7e250a
        exitCode: 0
        finishedAt: "2022-08-16T21:24:16Z"
        reason: Completed
        startedAt: "2022-08-16T21:24:16Z"
  hostIP: 10.0.113.252
  initContainerStatuses:
  - containerID: containerd://e25ffce27e57e170ac19cb12c390dce72f92400f04603d57ba2e54757dbbcc84
    image: docker.io/faiq/gpu-operator-validator:v1.11.0
    imageID: docker.io/faiq/gpu-operator-validator@sha256:0ec2a74daaaea3717a95d9399cbc6e9486c38f60e9b54b826c7031e3c2d52dfa
    lastState: {}
    name: cuda-validation
    ready: true
    restartCount: 0
    state:
      terminated:
        containerID: containerd://e25ffce27e57e170ac19cb12c390dce72f92400f04603d57ba2e54757dbbcc84
        exitCode: 0
        finishedAt: "2022-08-16T21:24:16Z"
        reason: Completed
        startedAt: "2022-08-16T21:24:15Z"
  phase: Succeeded
  podIP: 192.168.20.19
  podIPs:
  - ip: 192.168.20.19
  qosClass: BestEffort
  startTime: "2022-08-16T21:24:14Z"
```

Removing the compat libs worked 

```
$ docker run --entrypoint=/bin/bash -it faiq/gpu-operator-validator:v1.11.0 
[root@eb0f8ba184e2 /]# ls -R /usr/local/cuda-11.6/
/usr/local/cuda-11.6/:
cuda-11.6  lib64  targets

/usr/local/cuda-11.6/targets:
x86_64-linux

/usr/local/cuda-11.6/targets/x86_64-linux:
lib

/usr/local/cuda-11.6/targets/x86_64-linux/lib:
libOpenCL.so.1	libOpenCL.so.1.0  libOpenCL.so.1.0.0  libcudart.so.11.0  libcudart.so.11.6.55
[root@eb0f8ba184e2 /]# 
``` 

When would the next timeline for release be?

**Comment by shivamerla** [comment]
@faiq this fix will be part of our next operator release during late September.

**Comment by shivamerla** [comment]
closing this as the fix has been merged into gpu-operator/validator to remove cuda compat libraries. It will be part of next release (v22.09).

**Comment by cdesiniotis** [comment]
GPU Operator v22.9.0 has been released and contains the fix for this issue: https://github.com/NVIDIA/gpu-operator/releases/tag/v22.9.0
